### PR TITLE
First draft of data science

### PR DESCRIPTION
**Pull request by Tornquist**

### CODE REVIEW

**Comment by katieschilling** [comment]
I like to go through PRs by commit. Next time, separate each atomic change into its own commit so I can see the history of your project. This is especially important for larger projects that need to be merged to shared boxes.

File: data_science/data_science.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      unless @cohorts[line[:cohort]]
+        @cohorts[line[:cohort]] = Cohort.new(line[:cohort])
+      end
+
+      @cohorts[line[:cohort]].insert_record(line)
+    end
+  end
+
+  def symbolize_keys(hash)
+    result = {}
```
Comment (katieschilling) [inline data_science/data_science.rb]: use `each_with_object`

File: data_science/cohort.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    validate_cohort(record)
+    record = convert_record(record)
+    if record[:result] == 1
+      @conversions = @conversions + 1
+    end
+    @records.push(record)
+  end
+
+  def conversion_rate
+    interval = ABAnalyzer.confidence_interval(@conversions, @records.size, 0.95)
```
Comment (katieschilling) [inline data_science/cohort.rb]: move confidence level to a constant

File: data_science/rspec/cohort_spec.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    it "validates cohort name" do
+      expect { cohort.insert_record(wrong_record) }.to raise_error(NameError)
+    end
+  end
+
+  context "math check" do
+    let(:name) { "B" }
+    subject(:cohort) do
+      cohort = Cohort.new(name)
+      rates = [1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1]
```
Comment (katieschilling) [inline data_science/rspec/cohort_spec.rb]: rate is unclear

File: data_science/rspec/file_reader_spec.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  context 'negative' do
+    let(:file_path_nil) { nil }
+    let(:file_path_invalid) { 'file.file' }
+
+    it "should throw an error on a nil file path" do
+      expect { FileReader.new(file_path_nil) }.to raise_error(IOError)
+    end
+
```
Comment (katieschilling) [inline data_science/rspec/file_reader_spec.rb]: add test for valid json format output

Comment (Tornquist) [inline data_science/rspec/file_reader_spec.rb]: Test added in happy path. This validates the processing of JSON.parse

**Comment by Tornquist** [comment]
The updated cohort_spec uses the top level DataScience class to generate test data.  This is a test of  cohort methods that uses full integration to setup the seed data for the math operations.  I'm not completely satisfied with the location or the fact that DataScience was needed.  This change removes the dependencies on creating objects by hand within the test.

**Pull request closed without merging**